File: ts-core/src/net/node/destination.ts
import { ClusterId, NodeId } from "../node";
import { TransformSerdeable, TupleSerdeable } from "@core/serde";
import { OptionalClusterId } from "./clusterId";
import { UniqueKey } from "@core/object";

export class Destination implements UniqueKey {
    nodeId: NodeId;
    clusterId: OptionalClusterId;

    constructor(args?: { nodeId?: NodeId; clusterId: OptionalClusterId }) {
        this.nodeId = args?.nodeId ?? NodeId.broadcast();
        this.clusterId = args?.clusterId ?? ClusterId.noCluster();
    }

    static broadcast(): Destination {
        return new Destination();
    }

    static fromNodeId(nodeId: NodeId): Destination {
        return new Destination({ nodeId, clusterId: ClusterId.noCluster() });
    }

    isBroadcast(): boolean {
        return this.nodeId.isBroadcast() && this.clusterId.isNoCluster();
    }

    isUnicast(): boolean {
        return !this.nodeId.isBroadcast();
    }

    isMulticast(): boolean {
        return this.nodeId.isBroadcast() && !this.clusterId.isNoCluster();
    }

    equals(other: Destination): boolean {
        return this.nodeId.equals(other.nodeId) && this.clusterId.equals(other.clusterId);
    }

    static readonly serdeable = new TransformSerdeable(
        new TupleSerdeable([NodeId.serdeable, ClusterId.serdeable] as const),
        ([nodeId, clusterId]) => new Destination({ nodeId, clusterId }),
        (destination) => [destination.nodeId, destination.clusterId] as const,
    );

    toString(): string {
        return `Destination(${this.nodeId}, ${this.clusterId})`;
    }

    display(): string {
        return `${this.nodeId.display()} ${this.clusterId.display()}`;
    }

    uniqueKey(): string {
        return `${this.nodeId.uniqueKey()}+${this.clusterId.uniqueKey()}`;
    }

    toJSON(): string {
        return this.toString();
    }
}

File: py-core/py_core/net/node/destination.py
from dataclasses import dataclass
from typing import Self
from py_core.net.node.cluster_id import ClusterId

from py_core.net.node.node_id import NodeId
from py_core.serde.derive import DeriveDeserialize, DeriveSerialize


@dataclass(frozen=True)
class Destination(DeriveDeserialize, DeriveSerialize):
    node_id: NodeId
    cluster_id: ClusterId

    @classmethod
    def broadcast(cls) -> Self:
        return cls(NodeId.broadcast(), ClusterId.no_cluster())

    def is_broadcast(self) -> bool:
        return self.node_id.is_broadcast() and self.cluster_id.is_no_cluster()

    def is_multicast(self) -> bool:
        return self.node_id.is_broadcast() and not self.cluster_id.is_no_cluster()

    def is_unicast(self) -> bool:
        return not self.node_id.is_broadcast()

    @classmethod
    def loopback(cls) -> Self:
        return cls(NodeId.loopback(), ClusterId.no_cluster())

    def is_loopback(self) -> bool:
        return self.node_id.is_loopback()
